contracts/VaultV3.vy

```diff
@@ -1,135 +1,335 @@
+# @version 0.3.3
+
 from vyper.interfaces import ERC20
 
-# implements: ERC20
 # TODO: external contract: factory
 # TODO: external contract: access control
 # TODO: external contract: fee manager
 # TODO: external contract: healtcheck
 
 # INTERFACES #
-interface DetailedERC20:
-   def name() -> String[42]: view
-   def symbol() -> String[20]: view
-   def decimals() -> uint256: view
+interface ERC20Metadata:
+    def decimals() -> uint8: view
 
-interface Strategy:
+interface IStrategy:
    def asset() -> address: view
-
+   def vault() -> address: view
+   
 # EVENTS #
-event Transfer: 
-   sender: indexed(address)
-   receiver: indexed(address)
-   value: uint256
+event Transfer:
+    sender: indexed(address)
+    receiver: indexed(address)
+    value: uint256
+
+event Deposit:
+    recipient: indexed(address)
+    shares: uint256
+    amount: uint256
+
+event Withdraw:
+    recipient: indexed(address)
+    shares: uint256
+    amount: uint256
+
+event StrategyAdded: 
+   strategy: indexed(address)
+
+event StrategyRevoked: 
+   strategy: indexed(address)
+
+# STRUCTS #
+# TODO: strategy params
+struct StrategyParams:
+   activation: uint256
+   currentDebt: uint256
+   maxDebt: uint256
 
 # CONSTANTS #
 MAX_BPS: constant(uint256) = 10_000
 
+# STORAGE #
+asset: public(ERC20)
+strategies: public(HashMap[address, StrategyParams])
+balanceOf: public(HashMap[address, uint256])
+decimals: public(uint256)
+totalSupply: public(uint256)
+totalDebt: public(uint256)
+totalIdle: public(uint256)
 
-# # STORAGE #
-# strategies: public(HashMap[address, StrategyParams])
+@external
+def __init__(asset: ERC20):
+    self.asset = asset
+    self.decimals = convert(ERC20Metadata(asset.address).decimals(), uint256)
+    # TODO: implement
+    return
 
-#STRUCTS
-# TODO: strategy params
+# SUPPORT FUNCTIONS #
+@view
+@internal
+def _totalAssets() -> uint256:
+    return self.totalIdle + self.totalDebt
+
+@internal
+def _burnShares(shares: uint256, owner: address):
+    # TODO: do we need to check?
+    self.balanceOf[owner] -= shares
+    self.totalSupply -= shares
+
+@view
+@internal
+def _amountForShares(shares: uint256) -> uint256:
+    _totalSupply: uint256 = self.totalSupply
+    amount: uint256 = shares
+    if _totalSupply > 0:
+        amount = shares * self._totalAssets() / self.totalSupply
```

Comment:
We must change the way the price is calculated to prevent:
- Airdrop manipulation (hacks)
- First depositor attack manipulation